docker/ubuntu_18.04-dev.dockerfile

```diff
@@ -3,37 +3,64 @@ FROM ubuntu:18.04
 # Ubuntu Setup
 ENV TZ=America/Chicago
 RUN ln -snf /usr/share/zoneinfo/$TZ /etc/localtime && echo $TZ > /etc/timezone
+ARG PY_VERSION=2.7
 
 ENV HOME /root
-RUN apt-get update \
+RUN if [ "${PY_VERSION%.?}" -eq 3 ] ; \
+    then \ 
+            export PY_SUFIX=${PY_VERSION%.?}; \
+    fi;\
+    apt-get update \
     && apt-get install -y --fix-missing \
-            software-properties-common wget g++ \
-            build-essential python-numpy python-scipy cython python-setuptools \
-            python-nose git cmake vim emacs gfortran libblas-dev \
-            liblapack-dev libhdf5-dev libhdf5-serial-dev gfortran python-tables \
-            python-matplotlib python-jinja2 python-dev libpython-dev \
-            autoconf libtool python-setuptools python-pip doxygen \
-    && apt-get clean -y \
-    && pip install --force-reinstall \
+            software-properties-common \
+            wget \
+            g++ \
+            build-essential \
+            python${PY_SUFIX}-setuptools \
+            python${PY_SUFIX}-pip \
+            python${PY_SUFIX}-setuptools \
+            python${PY_SUFIX}-dev \
+            libpython${PY_SUFIX}-dev \
+            gfortran \
+            git \
+            cmake \
+            gfortran \
+            vim emacs \
+            libblas-dev \
+            liblapack-dev \
+            libhdf5-dev \
+            libhdf5-serial-dev \
+            autoconf \
+            libtool \
+            doxygen \
+    && apt-get clean -y; \
+    if [ "${PY_VERSION%.?}" -eq 3 ] ; \
+       then \ 
+            update-alternatives --install /usr/bin/python python /usr/bin/python3 10; \
+            update-alternatives --install /usr/bin/pip pip /usr/bin/pip3 10; \
+    fi;\
+    pip install --force-reinstall \
             sphinx \
             cloud_sptheme \
+            tables \
+            matplotlib \
+            jinja2 \
             prettytable \
             sphinxcontrib_bibtex \
             numpydoc \
             nbconvert \
             numpy \
+            scipy \
             cython
 
 # Script conditional setup: Default PyNE alone
 ARG build_moab=NO
 ARG enable_pymoab=NO
 ARG build_dagmc=NO
-ARG build_pyne=YES
 RUN echo "Configuration: \n" \
          "MOAB: $build_moab\n" \
          "pyMOAB: $enable_pymoab\n" \
-         "DAGMC: $build_dagmc\n" \
-         "PyNE: $build_pyne\n"
+         "DAGMC: $build_dagmc\n" 
```

Comment:
Spreading the declaration of those variable just before their use would allow to accelerate the docker generation process...
The issue is, that we will not be able to check the consistency of the argument in the dockerfile. It is checked in the python script, the dockerfile will not check it self consistent...

Reply:
I actually did it anyway as the python script check the options consistency